akka-http-core/src/main/scala/akka/http/impl/engine/rendering/HttpRequestRendererFactory.scala

```diff
@@ -99,7 +99,7 @@ private[http] class HttpRequestRendererFactory(userAgentHeader: Option[headers.`
       }
 
     def renderContentLength(contentLength: Long) =
-      if (method.isEntityAccepted) r ~~ `Content-Length` ~~ contentLength ~~ CrLf else r
+      if (method.isEntityAccepted && (contentLength > 0 || !method.isSafe)) r ~~ `Content-Length` ~~ contentLength ~~ CrLf else r
```

Comment:
Won't this line end up rendering all normal (i.e. empty) DELETE requests with a `Content-Length: 0`?
I think @2beaucoup is right, we now have essentially 3 types of HTTP methods:

1. The one where entities are *regular* (OPTIONS, PATCH, POST, PUT)
2. The one where entities are *disallowed* (CONNECT, HEAD, TRACE)
3. The ones where entities are *tolerated* (DELETE, GET)

We accept entities on request with methods whose `requestEntities` property is *regular* or *tolerated*, but we only render `Content-Length: 0` headers for the former type.

So, I'd propose we do the following:

1. Replace the `isEntityAccepted` property with one named `requestEntities`.

2. Introduce a simple ADT:
    ```scala
    sealed trait RequestEntityAcceptance
    object RequestEntityAcceptance {
      case object Regular extends RequestEntityAcceptance
      case object Tolerated extends RequestEntityAcceptance
      case object Disallowed extends RequestEntityAcceptance
    }
    ```

3. Change the rendering logic to only render `Content-Length: 0` headers for requests with `Regular` methods.

WDYT?